i18n - initial setup

mraveux opened the pull request:
This PR add to the already existing i18n:
- transifex support
- french language
- `lang` cookie detection, fallback to browser language
- the `src/translations/index.js`file is now `.gitignore`ed because it's auto-generated by the `src/translations/i18n.sh` script executed by the `prebuild` npm script.
- npm scripts:
  - `i18n:pull` to retrieve translations from transifex (only languages min 95% translated)
  - `i18n:push` to push the source file (`src/translations/en.json`) to transifex

One question: Do you prefer translation files to be consistent between all applications (`.po` files) or the `.json` version of the Keyguard is enough (current version)?

mraveux committed:
i18n - readme changes

mraveux committed:
i18n - initial setup

sisou commented on src/lib/I18n.js:
```diff
@@ -20,8 +20,37 @@ class I18n { // eslint-disable-line no-unused-vars
         this._fallbackLanguage = fallbackLanguage;
 
         // Re-enable next line (and remove line after) to enable internationalisation support
```
This comment can now be removed :wink:

sisou commented on src/lib/I18n.js:
```diff
@@ -20,8 +20,37 @@ class I18n { // eslint-disable-line no-unused-vars
+        const cookieLang = cookieMatch && decodeURIComponent(cookieMatch[2]);
+        const browserLang = navigator.language.split('-')[0];
+
+        return cookieLang || browserLang;
```
Calling `split` on a string does not take many CPU cycles, but it's still more work than _not_ doing it. So I suggest to only do it when actually necessary.

```suggestion

        return cookieLang || navigator.language.split('-')[0];
```

sisou commented on package.json:
```diff
@@ -13,7 +15,10 @@
     "watchfix": "tsc-watch --pretty --onSuccess \"yarn lintfix\"",
-    "postinstall": "cd client && yarn"
+    "postinstall": "cd client && yarn",
+    "i18n:pull": "tx pull --all --parallel --no-interactive --force --minimum-perc 95",
```
Instead of setting the minimum percentage here, why not set it in the .tx/config file (where it's currently set to `0`)?

mraveux replied:
It was just easier for testing to not set it in the .tx/config file. I change it now.

sisou commented on src/translations/de.json:
```diff
@@ -0,0 +1,165 @@
+{
+    "_language": "Deutsch",
+
+    "language-changed": "The display language changed but some translations can't be updated. Do you want to reload the page to update thoses?",
```
```suggestion
    "language-changed": "Die Anzeigesprache hat sich geändert, aber manche Übersetzungen können nicht aktualisiert werden. Möchtest du die Seite neu laden um auch diese zu aktualisieren?"
```

sisou commented on src/translations/i18n.sh:
```diff
@@ -0,0 +1,19 @@
+#!/bin/bash
+
+FICHIERS=$(ls -1 src/translations/*.json)
```
What is `FICHIERS`? Is that french for `FILES`?

mraveux replied:
oh, yes. 😅

sisou commented on src/translations/i18n.sh:
```diff
@@ -0,0 +1,19 @@
+
+for fichier in $FICHIERS; do
+    basename=${fichier##*/}
+    basename=${basename%.json}
```
What kind of formatting is going on here? Maybe explain it in a comment?

mraveux replied:
it's retrieving the basename without the file extension from the file path.
For example: `de` from `src/translations/de.json`.
I'll add a comment 👍

sisou replied:
But, then you are adding `.json` again, so it's useless? Or don't I understand?

mraveux replied:
starting from `src/translations/de.json`.
then basename = `de.json`
then basename = `de`
So i'm removing `.json`, not adding it 😉

mraveux committed:
i18n - small fixes

- set transifex cli minimum_perc arg in the .tx/config file
- minor optimisation in detectLanguage function
- add a translation for the ‘language-changed’ key in Deutsch
- rename variables in french to english & add comments in i18n.sh

danimoh committed:
Readme: update, extend for i18n and local testing

danimoh committed:
Internationalization: some improvements and fixes

- only return supported languages in detecLanguage
- temporarily disable defaulting to the browser's language, until
  language switching is enabled in the wallet
- change message on focus after background language change to the same
  one as in Hub.
- move i18n.sh to tools
- exit on first error in i18n.sh
- ensure consistent sorting of i18n files despite the locale or the
  operating system for deterministic builds (cf. 508871b)
- add i18n:build-dictionary npm script for invoking i18n.sh
- run i18n:build-dictionary before checklangs
- change a path in .gitignore for consistency

danimoh committed:
Internationalization: Add Chinese translation and update others

danimoh committed:
Internationalization: Add missing translations

Note that <([^>/](?!data-i18n="[^"]{1,100}"))+?>[^<]+?</ is a useful regex
for searching missing translations again in the future.

mraveux committed:
Update EN & FR translations

danimoh committed:
Internationalization: add i18n for PR #400 after rebase on master
